x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java

```diff
@@ -93,34 +105,126 @@ protected void readInternal(byte[] b, int offset, int length) throws IOException
         }
     }
 
-    private void readInternalBytes(final long part, final long pos, byte[] b, int offset, int length) throws IOException {
+    private void readInternalBytes(final int part, long pos, final byte[] b, int offset, int length) throws IOException {
+        final long currentSequentialReadSize = sequentialReadSize;
+        if (currentSequentialReadSize != NO_SEQUENTIAL_READ_OPTIMIZATION) {
+            final StreamForSequentialReads streamForSequentialReads = streamForSequentialReadsRef.get();
+            if (streamForSequentialReads == null) {
+                // start a new sequential read
+                if (tryReadAndKeepStreamOpen(part, pos, b, offset, length, currentSequentialReadSize)) {
+                    return;
+                }
+            } else if (streamForSequentialReads.part == part && streamForSequentialReads.pos == pos) {
+                // continuing a sequential read that we started previously
+                assert streamForSequentialReads.isFullyRead() == false;
+                int read = streamForSequentialReads.inputStream.read(b, offset, length);
+                assert read <= length : read + " vs " + length;
+                streamForSequentialReads.pos += read;
+                position += read;
+                pos += read;
+                offset += read;
+                length -= read;
+
+                if (streamForSequentialReads.isFullyRead()) {
+                    if (streamForSequentialReadsRef.compareAndSet(streamForSequentialReads, null) != false) {
+                        streamForSequentialReads.close();
+                    } else {
+                        // something happened concurrently, defensively stop optimizing
+                        sequentialReadSize = NO_SEQUENTIAL_READ_OPTIMIZATION;
+                    }
+
+                    if (length == 0) {
+                        // the current stream contained precisely enough data for this read, so we're good.
+                        return;
+                    } else {
+                        // the current stream didn't contain enough data for this read, so we must read more
+                        if (sequentialReadSize != NO_SEQUENTIAL_READ_OPTIMIZATION
```

Comment:
I guess this is to handle some kind of concurrency. I don't really understand the concurrency here though.

Reply:
Just closing.